web/src/client/http.js

```diff
@@ -38,11 +55,65 @@ class WSClient {
    * @param {URL} url - Websocket URL.
    */
   constructor(url) {
-    this.client = new WebSocket(url.toString());
+    this.url = url.toString();
+
+    this.event_handlers = [];
+    this.close_handlers = [];
+    this.error_handlers = [];
+    this.open_handlers = [];
+    this.reconnectAttempts = 0;
+    this.connect();
+  }
+
+  wsState() {
+    const state = this.client.readyState;
+    if ((state !== SocketStates.CONNECTED) && (this.reconnectAttempts >= MAX_ATTEMPTS)) return SocketStates.UNRECOVERABLE;
+
+    return state;
+  }
+
+  isRecoverable() {
+    return (this.wsState() !== SocketStates.UNRECOVERABLE);
+  }
+
+  isConnected() {
+    return (this.wsState() === SocketStates.CONNECTED);
+  }
+
+  subscribe_to_events() {
+    this.client.onopen = () => {
+      console.log("Connected websocket");
+      this.reconnectAttempts = 0;
+      if (this.timeout !== undefined) clearTimeout(this.timeout);
```

Comment:
I think that condition is not needed. See https://developer.mozilla.org/en-US/docs/Web/API/clearTimeout Just call it always